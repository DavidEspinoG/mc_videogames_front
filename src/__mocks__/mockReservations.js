export const mockReservation = {
  id: 26,
  user_id: 7,
  videogame_id: 68,
  days: 11,
  total_price: '220.0',
  created_at: '2023-07-20T17:53:13.512Z',
  updated_at: '2023-07-20T17:53:13.512Z',
  videogame: {
    id: 68,
    name: 'Street Fighter',
    photo: 'https://mms.businesswire.com/media/20230122005013/en/1692292/5/p1-1.jpg',
    description:
      'Street Fighter is a 1987 arcade game developed by Capcom. It is the first competitive fighting game produced by the company and the inaugural game in the Street Fighter series.',
    price_per_day: '20.0',
    created_at: '2023-07-20T17:37:47.367Z',
    updated_at: '2023-07-20T17:37:47.367Z',
  },
};

const mockReservations = [
  mockReservation,
  {
    id: 19,
    user_id: 7,
    videogame_id: 62,
    days: 2,
    total_price: '10.0',
    created_at: '2023-07-20T17:37:47.766Z',
    updated_at: '2023-07-20T17:37:47.766Z',
    videogame: {
      id: 62,
      name: 'Tetris',
      photo:
        'https://upload.wikimedia.org/wikipedia/en/thumb/7/7d/Tetris_NES_cover_art.jpg/220px-Tetris_NES_cover_art.jpg',
      description:
        'Tetris is a puzzle video game created in 1984 by Alexey Pajitnov, a Soviet software engineer.',
      price_per_day: '5.0',
      created_at: '2023-07-20T17:37:47.068Z',
      updated_at: '2023-07-20T17:37:47.068Z',
    },
  },
];

export default mockReservations;
