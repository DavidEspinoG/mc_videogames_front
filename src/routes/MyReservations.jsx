import { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useMediaQuery } from 'usehooks-ts';
import Carousel from '../components/Carousel';
import Spinner from '../components/spinner';
import { getReservations } from '../redux/slices/reservationsSlice';
import { selectReservations, selectReservationsError } from '../redux/store';

const MyReservations = () => {
  const dispatch = useDispatch();
  const reservations = useSelector(selectReservations);
  const error = useSelector(selectReservationsError);
  const isDesktop = useMediaQuery('(min-width: 768px)');
  const isLargeDesktop = useMediaQuery('(min-width: 1200px)');
  const [page, setPage] = useState(0);
  let itemsPerPage = 1;
  if (isLargeDesktop) {
    itemsPerPage = 3;
  } else if (isDesktop) {
    itemsPerPage = 2;
  }

  useEffect(() => {
    dispatch(getReservations());
  }, [dispatch]);

  if (error) {
    return <h1>{error}</h1>;
  }

  if (!reservations) {
    return (
      <div className="d-flex justify-content-center align-items-center w-100 h-100">
        <Spinner />
      </div>
    );
  }

  if (!reservations?.length) {
    return <h1>There are no reservations in your name</h1>;
  }

  const slicedItems = reservations.slice(page, page + itemsPerPage);
  const disabledLeft = page === 0;
  const disabledRight = page + itemsPerPage >= reservations.length;

  return (
    <div className="h-100 text-center d-flex flex-column justify-content-center">
      <h1 className="fw-bolder mb-5 pb-4 h2">MY RESERVATIONS</h1>
      <Carousel
        items={slicedItems}
        page={page}
        setPage={setPage}
        disabledLeft={disabledLeft}
        disabledRight={disabledRight}
      />
    </div>
  );
};

export default MyReservations;
