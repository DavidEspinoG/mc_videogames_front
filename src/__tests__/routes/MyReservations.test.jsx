import MyReservations from '../../routes/MyReservations';
import mockReservations from '../../__mocks__/mockReservations';
import renderWithProviders from '../test-utils';

describe('MyReservations', () => {
  it('should render as expected', () => {
    const { container } = renderWithProviders(<MyReservations />, {
      preloadedState: {
        reservations: {
          reservations: mockReservations,
        },
      },
    });
    expect(container).toMatchSnapshot();
  });
});
