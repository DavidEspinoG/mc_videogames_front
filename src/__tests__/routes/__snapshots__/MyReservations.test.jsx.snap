// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyReservations > should render as expected 1`] = `
<div>
  <div
    class="h-100 text-center d-flex flex-column justify-content-center"
  >
    <h1
      class="fw-bolder mb-5 pb-4 h2"
    >
      MY RESERVATIONS
    </h1>
    <section
      class="row mx-0 g-0"
      id="carousel"
    >
      <div
        class="col-auto d-flex flex-column justify-content-center"
      >
        <button
          class="btn-carousel btn-left p-3"
          data-testid="btn-left"
          disabled=""
          type="button"
        >
          <svg
            class="fs-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 12.796V3.204L4.519 8 10 12.796zm-.659.753-5.48-4.796a1 1 0 0 1 0-1.506l5.48-4.796A1 1 0 0 1 11 3.204v9.592a1 1 0 0 1-1.659.753z"
            />
          </svg>
        </button>
      </div>
      <div
        class="col row mx-0 row-cols-1 row-cols-md-2 row-cols-xl-3"
      >
        <article
          class="col"
        >
          <a
            href="/details/68"
          >
            <div
              class="vh-25 mb-4"
            >
              <img
                alt="Street Fighter"
                class="h-100 img-fluid rounded aspect-ratio-1"
                src="https://mms.businesswire.com/media/20230122005013/en/1692292/5/p1-1.jpg"
              />
            </div>
            <h3
              class="h6 fw-bolder text-uppercase mb-4"
            >
              Street Fighter
            </h3>
          </a>
          <div
            class="text-muted small"
          >
            <p
              class="mb-0"
            >
              <span
                class="fw-bold"
              >
                Starting date:
              </span>
               
              Thursday, July 20, 2023
            </p>
            <p
              class="mb-0"
            >
              <span
                class="fw-bold"
              >
                Finishing date:
              </span>
               
              Monday, July 31, 2023
            </p>
            <p
              class="mb-0"
            >
              <span
                class="fw-bold"
              >
                Days:
              </span>
               
              11
            </p>
            <p
              class="mb-0"
            >
              <span
                class="fw-bold"
              >
                Total price:
              </span>
               
              US$Â 220
            </p>
            <p
              class="d-webkit-box clamp-3"
            >
              Street Fighter is a 1987 arcade game developed by Capcom. It is the first competitive fighting game produced by the company and the inaugural game in the Street Fighter series.
            </p>
          </div>
        </article>
      </div>
      <div
        class="col-auto d-flex flex-column justify-content-center"
      >
        <button
          class="btn-carousel btn-right p-3"
          data-testid="btn-right"
          type="button"
        >
          <svg
            class="fs-5"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 12.796V3.204L11.481 8 6 12.796zm.659.753 5.48-4.796a1 1 0 0 0 0-1.506L6.66 2.451C6.011 1.885 5 2.345 5 3.204v9.592a1 1 0 0 0 1.659.753z"
            />
          </svg>
        </button>
      </div>
    </section>
  </div>
</div>
`;
